core/discovery/api/fetcher.go

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package api
+
+import (
+	"fmt"
+	"time"
+
+	log "github.com/cihub/seelog"
+	"github.com/mysteriumnetwork/node/core/discovery"
+	"github.com/mysteriumnetwork/node/market"
+)
+
+const (
+	fetcherLogPrefix = "[proposal-fetcher-api] "
+)
+
+// FetchCallback does real fetch of proposals through Mysterium API
+type FetchCallback func() ([]market.ServiceProposal, error)
+
+// Fetcher represents async proposal fether from Mysterium API
+type Fetcher struct {
+	fetch         FetchCallback
+	fetchInterval time.Duration
+	fetchShutdown chan bool
+
+	proposalStorage *discovery.ProposalStorage
+	proposalChan    chan market.ServiceProposal
+}
+
+// NewFetcher create instance of Fetcher
+func NewFetcher(proposalsStorage *discovery.ProposalStorage, callback FetchCallback, interval time.Duration) *Fetcher {
+	return &Fetcher{
+		fetch:         callback,
+		fetchInterval: interval,
+
+		proposalStorage: proposalsStorage,
+	}
+}
+
+// Start begins fetching proposals to storage
+func (fetcher *Fetcher) Start() error {
+	if err := fetcher.fetchDo(); err != nil {
+		return err
+	}
+
+	fetcher.fetchShutdown = make(chan bool, 1)
+	go fetcher.fetchLoop()
+
+	return nil
+}
+
+// Stop ends fetching proposals to storage
+func (fetcher *Fetcher) Stop() {
+	fetcher.fetchShutdown <- true
+}
+
+// SubscribeProposals allows to subscribe all fetched proposals
+func (fetcher *Fetcher) SubscribeProposals(proposalsChan chan market.ServiceProposal) {
```

Comment:
you should allow multiple subscribers here. Create an internal storage for channels, and allow for adding/removing from it.